test/e2e/network/netpol/network_policy.go

```diff
@@ -0,0 +1,751 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package netpol
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/kubernetes/test/e2e/network"
+	utilnet "k8s.io/utils/net"
+
+	"github.com/onsi/ginkgo"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = network.SIGDescribe("Netpol [LinuxOnly]", func() {
+	f := framework.NewDefaultFramework("netpol")
+
+	addSCTPContainers := false
+	protocols := []v1.Protocol{v1.ProtocolTCP, v1.ProtocolUDP}
+	if addSCTPContainers {
+		protocols = append(protocols, v1.ProtocolSCTP)
+	}
+	namespacePrefix := "netpol-test"
+	nsX := fmt.Sprintf("%s-x", namespacePrefix)
+	nsY := fmt.Sprintf("%s-y", namespacePrefix)
+	nsZ := fmt.Sprintf("%s-z", namespacePrefix)
+	model := NewModel([]string{nsX, nsY, nsZ}, []string{"a", "b", "c"}, []int32{80, 81}, protocols)
```

Comment:
the framework creates namespaces to avoid overlapping

Reply:
This is a good point, and we can have those namespaces prefixed by the unique for-free namespace.

I want to add more context as to why there are multiple namespaces:

A key value of these new tests is that they create 3 namespaces each with 3 pods, and probe all combinations ( 9 pods x 9 pods = 81 data points ) -- including cross-namespace calls.

So I think we do need additional namespaces beyond the single unique one the tests give us for free: otherwise the information content and value of the tests is diminished.

Here's an example of a test run:

```
Oct 19 10:34:16.907: INFO: expected:

-		netpol-test-x/a	netpol-test-x/b	netpol-test-x/c	netpol-test-y/a	netpol-test-y/b	netpol-test-y/c	netpol-test-z/a	netpol-test-z/b	netpol-test-z/c
netpol-test-x/a	X		.		.		.		.		.		.		.		.	
netpol-test-x/b	X		.		.		.		.		.		.		.		.	
netpol-test-x/c	X		.		.		.		.		.		.		.		.	
netpol-test-y/a	.		.		.		.		.		.		.		.		.	
netpol-test-y/b	.		.		.		.		.		.		.		.		.	
netpol-test-y/c	.		.		.		.		.		.		.		.		.	
netpol-test-z/a	X		.		.		.		.		.		.		.		.	
netpol-test-z/b	X		.		.		.		.		.		.		.		.	
netpol-test-z/c	X		.		.		.		.		.		.		.		.	



Oct 19 10:34:16.907: INFO: observed:

-		netpol-test-x/a	netpol-test-x/b	netpol-test-x/c	netpol-test-y/a	netpol-test-y/b	netpol-test-y/c	netpol-test-z/a	netpol-test-z/b	netpol-test-z/c
netpol-test-x/a	X		.		.		.		.		.		.		.		.	
netpol-test-x/b	X		.		.		.		.		.		.		.		.	
netpol-test-x/c	X		.		.		.		.		.		.		.		.	
netpol-test-y/a	.		.		.		.		.		.		.		.		.	
netpol-test-y/b	.		.		.		.		.		.		.		.		.	
netpol-test-y/c	.		.		.		.		.		.		.		.		.	
netpol-test-z/a	X		.		.		.		.		.		.		.		.	
netpol-test-z/b	X		.		.		.		.		.		.		.		.	
netpol-test-z/c	X		.		.		.		.		.		.		.		.	



Oct 19 10:34:16.907: INFO: comparison:

-		netpol-test-x/a	netpol-test-x/b	netpol-test-x/c	netpol-test-y/a	netpol-test-y/b	netpol-test-y/c	netpol-test-z/a	netpol-test-z/b	netpol-test-z/c
netpol-test-x/a	.		.		.		.		.		.		.		.		.	
netpol-test-x/b	.		.		.		.		.		.		.		.		.	
netpol-test-x/c	.		.		.		.		.		.		.		.		.	
netpol-test-y/a	.		.		.		.		.		.		.		.		.	
netpol-test-y/b	.		.		.		.		.		.		.		.		.	
netpol-test-y/c	.		.		.		.		.		.		.		.		.	
netpol-test-z/a	.		.		.		.		.		.		.		.		.	
netpol-test-z/b	.		.		.		.		.		.		.		.		.	
netpol-test-z/c	.		.		.		.		.		.		.		.		.
```

These tests are based on "truth tables" that compare the *expected* and *actual* connectivity of each of the 81 pairs.  Since network policies live in namespaces, and peers can be selected by namespace, showing the connectivity of pods in other namespaces is key information to show whether a network policy is working as intended or not.

Reply:
add a comment with that please, don't worry about the size, that is hard to infer from the code and seems really cool ;)

Reply:
👍